File: 2023/08.17/정호조/RGB거리2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
 * 조건 1) 인접한 집은 색이 달라야함, 조건 2) 1번과 N번 집은 색이 달라야함
 * 1번집이 빨, 초, 파를 골랐을 경우를 나누면서 시작
 * 1번이 빨간색을 골랐다면 N번은 빨간색을 제외한 초, 파 중 골라야함. 즉 dp[N][color]의 값이 조건 2까지 부합하는 경우의 최소 비용
 * 크기가 3인 cases[] 배열에 1번 집이 각 색깔을 골랐을 때의 최소비용 저장 후 cases[] 배열의 최솟값을 구하면 정답
 * */
public class RGB거리2_17404 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());

        int[][] home = new int[N + 1][3];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < 3; j++) {
                home[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[] cases = new int[3];
        for (int color = 0; color < 3; color++) {
            int[][] dp = new int[N+1][3];
            
            for(int i=0; i<N; i++){
                for(int j=0; j<3; j++){
                    dp[i][j] = home[i][j];
                }
            }

            for (int i = 1; i <= N; i++) {
                for (int j = 0; j < 3; j++) {
                    if (i == 1) {
                        if (j != color) {
                            dp[i][j] += dp[i - 1][color];
                        } else {
                            dp[i][color] = Integer.MAX_VALUE;
                        }
                    } else {
                        dp[i][j] += Math.min(dp[i - 1][(j + 1) % 3], dp[i - 1][(j + 2) % 3]);
                    }
                }
            }

            cases[color] = dp[N][color];
        }

        int ans = Integer.MAX_VALUE;
        for(int i=0; i<3; i++){
            ans = Math.min(cases[i], ans);
        }

        System.out.println(ans);
    }
}

File: 2023/08.17/조수연/RGB거리_2.kt
import java.io.*
import java.util.*
import kotlin.math.min

lateinit var st: StringTokenizer
const val MAX_VALUE = 1000000

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val n = readLine().toInt()
    val house = Array(n + 1) { IntArray(n + 1) }
    val dp = Array(n + 1) { IntArray(n + 1) }

    repeat(n) {
        st = StringTokenizer(readLine())
        house[it + 1][0] = st.nextToken().toInt()
        house[it + 1][1] = st.nextToken().toInt()
        house[it + 1][2] = st.nextToken().toInt()
    }

    var answer = MAX_VALUE

    for (i in 0 until 3) { // R -> G -> B순으로 선택
        for (j in 0 until 3) { // 선택한 색깔 집을 제외하고는 최댓값으로 갱신
            if (i == j) dp[1][j] = house[1][j]
            else dp[1][j] = MAX_VALUE
        }

        for (j in 2..n){
            dp[j][0] = house[j][0] + min(dp[j-1][1], dp[j-1][2])
            dp[j][1] = house[j][1] + min(dp[j-1][0], dp[j-1][2])
            dp[j][2] = house[j][2] + min(dp[j-1][0], dp[j-1][1])
        }

        for (j in 0 until 3){
            if (i != j) answer = min(answer, dp[n][j]) // n번이 1번에서 선택한 집이 아닐 때 최솟값 갱신
        }
    }

    bw.write("$answer")

    bw.flush()
    bw.close()
}